arangod/V8Server/v8-voccursor.cpp

```diff
@@ -157,6 +164,386 @@ static void JS_JsonCursor(v8::FunctionCallbackInfo<v8::Value> const& args) {
   TRI_V8_TRY_CATCH_END
 }
 
+// .............................................................................
+// cursor v8 class
+// .............................................................................
+
+struct V8Cursor final {
+  static constexpr uint16_t CID = 455656;
+  
+  V8Cursor(v8::Isolate* isolate,
+           v8::Handle<v8::Object> holder,
+           TRI_vocbase_t& vocbase,
+           CursorId cursorId)
+  : _isolate(isolate),
+    _cursorId(cursorId),
+    _resolver(vocbase),
+    _cte(transaction::Context::createCustomTypeHandler(vocbase, _resolver)) {
+    // sanity checks
+    TRI_ASSERT(_handle.IsEmpty());
+    TRI_ASSERT(holder->InternalFieldCount() > 0);
+    
+    // create a new persistent handle
+    holder->SetAlignedPointerInInternalField(0, this);
+    _handle.Reset(_isolate, holder);
+    _handle.SetWrapperClassId(CID);
+    
+    // and make it weak, so that we can garbage collect
+    _handle.SetWeak(&_handle, weakCallback, v8::WeakCallbackType::kFinalizer);
+    _options.customTypeHandler = _cte.get();
+  }
+  
+  ~V8Cursor() {
+    if (!_handle.IsEmpty()) {
+      TRI_ASSERT(_handle.IsNearDeath());
+      
+      _handle.ClearWeak();
+      v8::Local<v8::Object> data =
+      v8::Local<v8::Object>::New(_isolate, _handle);
+      data->SetInternalField(0, v8::Undefined(_isolate));
+      _handle.Reset();
+    }
+    if (_isolate) {
+      TRI_GET_GLOBALS2(_isolate);
+      TRI_vocbase_t* vocbase = v8g->_vocbase;
+      if (vocbase) {
+        CursorRepository* cursors = vocbase->cursorRepository();
+        cursors->remove(_cursorId, Cursor::CURSOR_VPACK);
+      }
+    }
+  }
+  
+  //////////////////////////////////////////////////////////////////////////////
+  /// @brief unwraps a structure
+  //////////////////////////////////////////////////////////////////////////////
+  
+  static V8Cursor* unwrap(v8::Local<v8::Object> handle) {
+    TRI_ASSERT(handle->InternalFieldCount() > 0);
+    return static_cast<V8Cursor*>(handle->GetAlignedPointerFromInternalField(0));
+  }
+
+  /// @brief return false on error
+  bool maybeFetchBatch(v8::Isolate* isolate) {
+    if (_dataIterator == nullptr && _hasMore) { // fetch more data
+      TRI_GET_GLOBALS();
+      CursorRepository* cursors = v8g->_vocbase->cursorRepository();
+      bool busy;
+      Cursor* cc = cursors->find(_cursorId, Cursor::CURSOR_VPACK, busy);
+      if (busy || cc == nullptr) {
+        return true; // someone else is using it
+      }
+      TRI_DEFER(cc->release());
+      
+      Result r = fetchData(cc);
+      if (r.fail()) {
+        TRI_CreateErrorObject(isolate, r);
+        return false;
+      }
+    }
+    return true;
+  }
+  
+  /// @brief fetch the next batch
+  Result fetchData(Cursor* cursor) {
+    TRI_ASSERT(_hasMore);
+    TRI_ASSERT(_dataIterator == nullptr);
+    TRI_ASSERT(cursor != nullptr && cursor->isUsed());
+    
+    _tmpResult.clear();
```

Comment:
I think we should always reset `_extraSlice`, `_dataSlice` and `_dataIterator` here too.
Otherwise it is at least theoretically possible that these point into `_tmpResult`, which is invalidated here. And there is at least theoretically no guarantee that these values will be re-populated in lines 262-272. So for safety reasons we should first reset these values, so that accessing them later will never crash the program.